src/platform/Linux/CHIPBluezHelper.cpp

```diff
@@ -2008,6 +2008,104 @@ CHIP_ERROR ConnectDevice(BluezDevice1 * apDevice)
     return error;
 }
 
+AdapterIterator::~AdapterIterator()
+{
+    g_object_unref(mManager);
+    g_list_free_full(mObjectList, g_object_unref);
+}
+
+void AdapterIterator::Initialize()
+{
+    GError * error = nullptr;
+
+    mManager = g_dbus_object_manager_client_new_for_bus_sync(G_BUS_TYPE_SYSTEM, G_DBUS_OBJECT_MANAGER_CLIENT_FLAGS_NONE,
+                                                             BLUEZ_INTERFACE, "/", bluez_object_manager_client_get_proxy_type,
+                                                             nullptr /* unused user data in the Proxy Type Func */,
+                                                             nullptr /*destroy notify */, nullptr /* cancellable */, &error);
+
+    VerifyOrExit(mManager != nullptr, ChipLogError(DeviceLayer, "Failed to get DBUS object manager for listing adapters."));
+
+    mObjectList      = g_dbus_object_manager_get_objects(mManager);
+    mCurrentListItem = mObjectList;
+
+exit:
+    if (error != nullptr)
+    {
+        ChipLogError(DeviceLayer, "DBus error: %s", error->message);
+        g_error_free(error);
+    }
+}
+
+template <size_t N>
+void safeNullTerminatedCopy(char (&dest)[N], const char * src)
+{
+    strncpy(dest, src, N);
+    dest[N - 1] = 0;
+}
+
+void AdapterIterator::Advance()
+{
+    if (mCurrentListItem == nullptr)
+    {
+        return;
+    }
+
+    bool foundAdapter = false;
+
+    while (!foundAdapter && (mCurrentListItem != nullptr))
+    {
+
+        BluezObject * object = BLUEZ_OBJECT(mCurrentListItem->data);
+        GList * interfaces   = g_dbus_object_get_interfaces(G_DBUS_OBJECT(object));
+
+        for (GList * it = interfaces; it != nullptr; it = it->next)
+        {
+            if (!BLUEZ_IS_ADAPTER1(it->data))
+            {
+                continue;
+            }
+
+            BluezAdapter1 * adapter = BLUEZ_ADAPTER1(it->data);
```

Comment:
Wouldn't `bluez_object_get_adapter1(object)` be an equivalent solution which doesn't require the inner loop?

Reply:
Thanks! I was looking at existing code which was not using this method. Code is much cleaner now (will update soon once I address all other comments)